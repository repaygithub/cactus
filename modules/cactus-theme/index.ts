export interface CactusTheme {
  colors: {
    /** Core colors */
    base: string
    baseText: string
    callToAction: string
    callToActionText: string

    /** Contrasts */
    lightContrast: string
    mediumContrast: string
    darkContrast: string
    darkestContrast: string

    /** Neutrals */
    white: string
    lightGray: string
    mediumGray: string
    darkGray: string

    /** Notification Colors */
    success: string
    warning: string
    error: string
  }
}

export type CactusColor = keyof CactusTheme['colors']

export interface GeneratorOptions {
  primaryHue: number
}

const repayOptions: GeneratorOptions = { primaryHue: 200 }

export function generateTheme({ primaryHue }: GeneratorOptions = repayOptions): CactusTheme {
  return {
    colors: {
      /** Core colors */
      base: `hsl(${primaryHue}, 96%, 11%)`,
      baseText: `hsl(0, 0%, 100%)`,
      callToAction: `hsl(${primaryHue}, 96%, 35%)`,
      callToActionText: `hsl(0, 0%, 100%)`,

      /** Contrasts */
      lightContrast: `hsl(${primaryHue}, 29%, 90%)`,
      mediumContrast: `hsl(${primaryHue}, 18%, 80%)`,
      darkContrast: `hsl(${primaryHue}, 96%, 11%)`,
      darkestContrast: `hsl(${primaryHue}, 10%, 35%)`,

      /** Neutrals */
      white: `hsl(0, 0%, 100%)`,
      lightGray: `hsl(0, 0%, 90%)`,
      mediumGray: `hsl(0, 0%, 70%)`,
      darkGray: `hsl(0, 0%, 50%)`,

      /** Notification Colors */
      success: `hsl(145, 89%, 28%)`,
      error: `hsl(353, 84%, 44%)`,
      warning: `hsl(47, 82%, 47%)`,
    },
  }
}

export default generateTheme()
