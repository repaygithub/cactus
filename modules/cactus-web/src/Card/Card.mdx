---
name: Card
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import Card from './Card'
import cactusTheme from '@repay/cactus-theme'

# Card

## Basic usage

Cards are box containers used to organize content. Content and actions used for cards regard one topic. Cards can have any height or width.

```jsx
import React from 'react'
import { Card } from '@repay/cactus-web'
export default () => (
  <Card margin="40px">
    <h2 style={{ margin: 0 }}>Title</h2>
    <h4 style={{ margin: 0, fontWeight: 400, fontSize: '12px' }}>Subtitle</h4>

    <table style={{ fontSize: '15px', borderCollapse: 'collapse' }}>
      <tbody>
        <tr>
          <th style={{ textAlign: 'left', padding: '0 60px 0 0', fontSize: '12px' }}>
            Total Amount Due
          </th>
          <td>
            <span style={{ textAlign: 'right', fontWeight: 600, fontSize: '25.92px' }}>
              $350.20
            </span>
          </td>
        </tr>

        <tr>
          <th style={{ textAlign: 'left', fontWeight: 'normal', fontSize: '12px' }}> Due Date </th>
          <td style={{ textAlign: 'right', fontWeight: 600, fontSize: '12px' }}>
            {new Date().toLocaleDateString('en-US')}
          </td>
        </tr>

        <tr>
          <th style={{ textAlign: 'left', fontWeight: 'normal', fontSize: '12px' }}>
            Minimum Amount Due
          </th>
          <td style={{ textAlign: 'right', fontWeight: 600, fontSize: '12px' }}>$127.00</td>
        </tr>
      </tbody>
    </table>
  </Card>
)
```

## Properties

<PropsTable of={Card} />
