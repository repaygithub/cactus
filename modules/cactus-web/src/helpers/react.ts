import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
