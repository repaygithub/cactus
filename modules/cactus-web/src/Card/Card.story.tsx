import { date, text } from '@storybook/addon-knobs/react'
import { storiesOf } from '@storybook/react'
import Box from '../Box/Box'
import Card from './Card'
import React from 'react'

storiesOf('Card', module).add('Basic Usage', () => (
  <Card margin="40px">
    <h2 style={{ margin: 0 }}>{text('Title', 'Title')}</h2>
    <h4 style={{ margin: '0 0 8px', fontWeight: 400, fontSize: '12px' }}>
      {text('Subtitle', 'Subtitle')}
    </h4>

    <table style={{ fontSize: '15px', borderCollapse: 'collapse' }}>
      <tbody>
        <tr>
          <Box as="th" style={{ textAlign: 'left', padding: '4px 60px 4px 0', fontSize: '12px' }}>
            Total Amount Due
          </Box>
          <td>
            <span style={{ textAlign: 'right', fontWeight: 600, fontSize: '25.92px' }}>
              {text('Total Amount', '$350.20')}
            </span>
          </td>
        </tr>

        <tr>
          <th style={{ textAlign: 'left', fontWeight: 'normal', fontSize: '12px' }}> Due Date </th>
          <td style={{ textAlign: 'right', fontWeight: 600, fontSize: '12px' }}>
            {new Date(date('Date', new Date())).toLocaleDateString()}
          </td>
        </tr>

        <tr>
          <th style={{ textAlign: 'left', fontWeight: 'normal', fontSize: '12px' }}>
            Minimum Amount Due
          </th>
          <td style={{ textAlign: 'right', fontWeight: 600, fontSize: '12px' }}>
            {text('Minimum Amount', '$127.00')}
          </td>
        </tr>
      </tbody>
    </table>
  </Card>
))
