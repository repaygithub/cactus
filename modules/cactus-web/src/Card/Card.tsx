import { MarginProps, margins } from '../helpers/margins'
import styled from 'styled-components'

interface CardProps extends MarginProps {}

export const Card = styled.div<CardProps>`
  ${margins}
  ${p => p.theme.colorStyles.standard};
  border-radius: 8px;
  padding: ${p => p.theme.space[4]}px;
  box-shadow: 0 3px 6px 0 ${p => p.theme.colors.callToAction};

  :hover {
    box-shadow: 0 4px 8px 0 ${p => p.theme.colors.callToAction};
  }

  & > & {
    padding: ${p => p.theme.space[5]}px;
    color: pink;
  }
  & > & > & {
    padding: ${p => p.theme.space[6]}px;
    color: blue;
  }
`

export default Card
