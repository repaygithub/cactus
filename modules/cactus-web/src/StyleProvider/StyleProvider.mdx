---
name: StyleProvider
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import StyleProvider from './StyleProvider'
import cactusTheme from '@repay/cactus-theme'

# StyleProvider

The `StyleProvider` is a wrapper for `styled-components`' `ThemeProvider` component which gives you the option to include some global styles.

# Best practices

- Make sure that any components from `@repay/cactus-web` are children of `StyleProvider`. This will ensure that each component has the correct styles.

## Basic usage

```jsx
import React from 'react'
import { StyleProvider } from '@repay/cactus-web'
import cactusTheme from '@repay/cactus-theme'

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      // Include routes/rest of application here
    </StyleProvider>
  )
}
```

# Breakpoints

The `StyleProvider` has a media queries scale which is based on the provided breaking points and is added to the given theme. The media queries have screen sizes small, medium, large, and extraLarge. The set media queries will be larger than the default, for a mobile first approach.

Props can also vary depending on the current media query. This is done by setting a prop to an array, where the array holds different values for each media query size. The first element of the array maps to the default screen size, and each element after maps to the set media queries from smallest to largest. This is used below in the example code to set the background color.

For more information on Breakpoints and Media Queries visit: https://styled-system.com/theme-specification#breakpoints

## Media Queries Usage

```jsx
import React from 'react'
import { StyleProvider, Box } from '.@repay/cactus-web'
import styled from 'styled-components'

const BreakpointBox = styled(Box)`
  border-radius: 15%;

  ${p => p.theme.mediaQueries.small} {
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extraLarge} {
    border-radius: 50%;
  }
`
const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <BreakpointBox backgroundColor={backgroundColors} width="100px" height="100px" />
    </StyleProvider>
  )
}
```

# Properties

<PropsTable of={StyleProvider} staticProp="StyleProvider" />
