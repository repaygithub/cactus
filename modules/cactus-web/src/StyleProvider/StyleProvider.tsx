import cactusTheme, { CactusTheme } from '@repay/cactus-theme'
import PropTypes from 'prop-types'
import React from 'react'
import * as styledComponents from 'styled-components'

import { breakpointOrder, breakpoints } from '../helpers/constants'
import { textStyle } from '../helpers/theme'

interface Env {
  NODE_ENV: string
}

interface Process {
  env: Env
}

declare let process: Process

const { createGlobalStyle } = styledComponents as styledComponents.ThemedStyledComponentsModule<
  CactusTheme
>

const DebugStyle = createGlobalStyle`
  body {
    --reach-tooltip: 1;
  }

  :root {
    --reach-dialog: 1;
  }
`

const queries = {
  small: `@media screen and (min-width: ${breakpoints.small}px)`,
  medium: `@media screen and (min-width: ${breakpoints.medium}px)`,
  large: `@media screen and (min-width: ${breakpoints.large}px)`,
  extraLarge: `@media screen and (min-width: ${breakpoints.extraLarge}px)`,
}

const GlobalStyle = createGlobalStyle`
html,
body,
button,
input,
select,
textarea {
  font-family: ${(p) => p.theme.font};
  font-weight: 400;
  color: ${(p): string => p.theme.colors.darkestContrast};
  font-style: normal;
  font-stretch: normal;
  letter-spacing: normal;
}

/* Inexplicably, applying this to <button> causes integration tests to fail. */
html,
body,
input,
select,
textarea {
  ${(p) => textStyle(p.theme, 'body')};
}

html,
body {
  margin: 0;
}

small {
  ${(p) => textStyle(p.theme, 'small')}
}

h1 {
  ${(p) => textStyle(p.theme, 'h1')};
}

h2 {
  ${(p) => textStyle(p.theme, 'h2')};
}

h3 {
  ${(p) => textStyle(p.theme, 'h3')};
}

h4, h5, h6 {
  ${(p) => textStyle(p.theme, 'h4')};
}
`

let shouldCheckTheme = process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'test'

const checkThemeProperties = (theme: CactusTheme): void => {
  if (theme.colors.transparentCTA === undefined) {
    console.warn(
      "You are using an outdated version of @repay/cactus-theme. Some features won't be available in @repay/cactus-web with this version. Please upgrade to @repay/cactus-theme >= 0.4.3."
    )
  }
  shouldCheckTheme = false
}

interface StyleProviderProps extends Omit<styledComponents.ThemeProviderProps<any, any>, 'theme'> {
  theme?: CactusTheme
  global?: boolean
}

export const StyleProvider: React.FC<StyleProviderProps> = (props): React.ReactElement => {
  const { global, children, theme: providedTheme, ...themeProviderProps } = props
  shouldCheckTheme && checkThemeProperties(providedTheme || cactusTheme)
  const theme = providedTheme ? providedTheme : cactusTheme
  theme.breakpoints = breakpointOrder.map((bp): string => `${breakpoints[bp]}px`)
  theme.mediaQueries = queries

  return (
    <styledComponents.ThemeProvider theme={theme} {...themeProviderProps}>
      <React.Fragment>
        <DebugStyle />
        {global && <GlobalStyle />}
        {children}
      </React.Fragment>
    </styledComponents.ThemeProvider>
  )
}

// PropTypes declarations of some recurring types so they can be re-used.
const textStyleObj = PropTypes.shape({
  fontSize: PropTypes.string.isRequired,
  lineHeight: PropTypes.string.isRequired,
}).isRequired
const textStylesObj = PropTypes.shape({
  tiny: textStyleObj,
  small: textStyleObj,
  body: textStyleObj,
  h4: textStyleObj,
  h3: textStyleObj,
  h2: textStyleObj,
  h1: textStyleObj,
}).isRequired
const colorStyleObj = PropTypes.shape({
  color: PropTypes.string.isRequired,
  backgroundColor: PropTypes.string.isRequired,
}).isRequired
const statusObj = PropTypes.shape({
  success: PropTypes.string.isRequired,
  warning: PropTypes.string.isRequired,
  error: PropTypes.string.isRequired,
}).isRequired

const generateArrayValidator = (expectedLength: number) =>
  PropTypes.arrayOf<number>((propValue, key, componentName, _, propFullName) => {
    if (propValue.length !== expectedLength) {
      return new Error(
        `Invalid prop \`${propFullName}\` supplied to \`${componentName}\`. ${propFullName.substring(
          0,
          propFullName.length - 3
        )} must be an array of length ${expectedLength}, but it is length ${propValue.length}.`
      )
    }
    const type = typeof propValue[key]
    if (type !== 'number') {
      return new Error(
        `Invalid prop \`${propFullName}\` of type \`${type}\` supplied to \`${componentName}\`, expected \`number\`.`
      )
    }
    return null
  }).isRequired

StyleProvider.propTypes = {
  global: PropTypes.bool,
  // @ts-ignore
  theme: PropTypes.shape({
    colors: PropTypes.shape({
      base: PropTypes.string.isRequired,
      baseText: PropTypes.string.isRequired,
      callToAction: PropTypes.string.isRequired,
      callToActionText: PropTypes.string.isRequired,
      transparentCTA: PropTypes.string.isRequired,
      lightContrast: PropTypes.string.isRequired,
      mediumContrast: PropTypes.string.isRequired,
      darkContrast: PropTypes.string.isRequired,
      darkestContrast: PropTypes.string.isRequired,
      white: PropTypes.string.isRequired,
      lightGray: PropTypes.string.isRequired,
      mediumGray: PropTypes.string.isRequired,
      darkGray: PropTypes.string.isRequired,
      success: PropTypes.string.isRequired,
      warning: PropTypes.string.isRequired,
      error: PropTypes.string.isRequired,
      transparentSuccess: PropTypes.string.isRequired,
      transparentWarning: PropTypes.string.isRequired,
      transparentError: PropTypes.string.isRequired,
      errorDark: PropTypes.string.isRequired,
      warningDark: PropTypes.string.isRequired,
      successDark: PropTypes.string.isRequired,
      status: PropTypes.shape({
        background: statusObj,
        avatar: statusObj,
      }).isRequired,
    }).isRequired,
    space: generateArrayValidator(8),
    fontSizes: generateArrayValidator(7),
    mobileFontSizes: generateArrayValidator(7),
    iconSizes: generateArrayValidator(4),
    textStyles: textStylesObj,
    mobileTextStyles: textStylesObj,
    colorStyles: PropTypes.shape({
      base: colorStyleObj,
      callToAction: colorStyleObj,
      standard: colorStyleObj,
      lightContrast: colorStyleObj,
      darkestContrast: colorStyleObj,
      success: colorStyleObj,
      error: colorStyleObj,
      warning: colorStyleObj,
      disable: colorStyleObj,
      transparentCTA: colorStyleObj,
      transparentError: colorStyleObj,
      transparentSuccess: colorStyleObj,
      transparentWarning: colorStyleObj,
      errorDark: colorStyleObj,
      warningDark: colorStyleObj,
      successDark: colorStyleObj,
    }).isRequired,
    border: PropTypes.oneOf(['thin', 'thick']).isRequired,
    shape: PropTypes.oneOf(['square', 'intermediate', 'round']).isRequired,
    font: PropTypes.string.isRequired,
    boxShadows: PropTypes.bool.isRequired,
  }),
}

StyleProvider.defaultProps = {
  global: false,
}

export default StyleProvider
