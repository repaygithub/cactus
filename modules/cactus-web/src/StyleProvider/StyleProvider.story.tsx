import React from 'react'

import { storiesOf } from '@storybook/react'
import Box from '../Box/Box'
import cactusTheme from '@repay/cactus-theme'
import Grid from '../Grid/Grid'
import styled from 'styled-components'
import StyleProvider from './StyleProvider'

const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const BreakpointBox = styled(Box)`
  max-width: 320px;
  width: 50px;
  height: 50px;
  border-radius: 15%

  ${p => p.theme.mediaQueries.small} {
    max-width: 708px;
    width: 75px;
    height 75px;
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    max-width: 964px;
    width: 100px;
    height: 100px;
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    max-width: 1140px;
    width: 150px;
    height: 150px;
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extralarge} {
    max-width: 1380px;
    width: 200px;
    height: 200px;
    border-radius: 50%;
  }
`

storiesOf('Style Provider', module).add('Component Adjusments based on Media Queries', () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <Grid justify="center">
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
      </Grid>
    </StyleProvider>
  )
})
