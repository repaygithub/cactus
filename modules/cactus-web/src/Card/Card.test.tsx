import { generateTheme } from '@repay/cactus-theme'
import { render } from '@testing-library/react'
import * as React from 'react'

import { StyleProvider } from '../StyleProvider/StyleProvider'
import Card from './Card'

describe('component: Card', (): void => {
  test('snapshot', (): void => {
    const { container } = render(
      <StyleProvider>
        <Card />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })
  test('should support margin space props', (): void => {
    const { container } = render(
      <StyleProvider>
        <Card m={2}>Content</Card>
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  describe('with theme customization', (): void => {
    test('should have no box shadow & 2px borders', (): void => {
      const theme = generateTheme({ primaryHue: 200, boxShadows: false, border: 'thick' })
      const { asFragment } = render(
        <StyleProvider theme={theme}>
          <Card />
        </StyleProvider>
      )

      expect(asFragment()).toMatchSnapshot()
    })

    test('should have 4px border radius', (): void => {
      const theme = generateTheme({ primaryHue: 200, shape: 'intermediate' })
      const { asFragment } = render(
        <StyleProvider theme={theme}>
          <Card />
        </StyleProvider>
      )

      expect(asFragment()).toMatchSnapshot()
    })

    test('should have 1px border radius', (): void => {
      const theme = generateTheme({ primaryHue: 200, shape: 'square' })
      const { asFragment } = render(
        <StyleProvider theme={theme}>
          <Card />
        </StyleProvider>
      )

      expect(asFragment()).toMatchSnapshot()
    })
  })
})
