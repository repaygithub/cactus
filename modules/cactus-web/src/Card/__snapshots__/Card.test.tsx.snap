// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Card should support margin space props 1`] = `
.c2 > .c0 {
  padding: 24px;
  color: pink;
}

.c2 > .c0 > .c0 {
  padding: 32px;
  color: blue;
}

.c1 {
  margin: 4px;
  background-color: hsl(0,0%,100%);
  color: hsl(200,10%,20%);
  border-radius: 8px;
  padding: 16px;
  box-shadow: 0 3px 6px 0 hsl(200,96%,35%);
}

.c1:hover {
  box-shadow: 0 4px 8px 0 hsl(200,96%,35%);
}

.c1 > .c0 {
  padding: 24px;
  color: pink;
}

.c1 > .c0 > .c0 {
  padding: 32px;
  color: blue;
}

<div>
  <div
    class="c0 c1"
  >
    Content
  </div>
</div>
`;

exports[`component: Card snapshot 1`] = `
.c1 {
  background-color: hsl(0,0%,100%);
  color: hsl(200,10%,20%);
  border-radius: 8px;
  padding: 16px;
  box-shadow: 0 3px 6px 0 hsl(200,96%,35%);
}

.c1:hover {
  box-shadow: 0 4px 8px 0 hsl(200,96%,35%);
}

.c1 > .c0 {
  padding: 24px;
  color: pink;
}

.c1 > .c0 > .c0 {
  padding: 32px;
  color: blue;
}

<div>
  <div
    class="c0 c1"
  />
</div>
`;
